src/backends/webgl.ts

```diff
@@ -311,24 +349,177 @@ let compile_rules: ASTVisit<Emitter, string> =
+        let op: string = (tree.op === "+" ? "add" : "subtract");
+        if (typ === FLOAT2) {
+          if (typL === FLOAT2 && typR === FLOAT2) {
+            return `vec2.${op}(vec2.create(), ${lhs}, ${rhs})`;
+          } else if (typL === FLOAT2 && (typR === FLOAT || typR === INT)) {
+            return `vec2.${op}(vec2.create(), ${lhs}, vec2.fromValues(${rhs}, ${rhs}))`;
+          } else if ((typL === FLOAT || typL === INT) && typR === FLOAT2) {
+            return `vec2.${op}(vec2.create(), vec2.fromValues(${lhs}, ${lhs}), ${rhs})`;
+          }
+        } else if (typ === FLOAT3) {
+          if (typL === FLOAT3 && typR === FLOAT3) {
+            return `vec3.${op}(vec3.create(), ${lhs}, ${rhs})`;
+          } else if (typL === FLOAT3 && (typR === FLOAT || typR === INT)) {
+            return `vec3.${op}(vec3.create(), ${lhs}, vec3.fromValues(${rhs}, ${rhs}, ${rhs}))`;
+          } else if ((typL === FLOAT || typL === INT) && typR === FLOAT3) {
+            return `vec3.${op}(vec3.create(), vec3.fromValues(${lhs}, ${lhs}, ${lhs}), ${rhs})`;
+          }
+        } else if (typ === FLOAT4) {
+          if (typL === FLOAT4 && typR === FLOAT4) {
+            return `vec4.${op}(vec4.create(), ${lhs}, ${rhs})`;
+          } else if (typL === FLOAT4 && (typR === FLOAT || typR === INT)) {
+            return `vec4.${op}(vec4.create(), ${lhs}, vec4.fromValues(${rhs}, ${rhs}, ${rhs}, ${rhs}))`;
+          } else if ((typL === FLOAT || typL === INT) && typR === FLOAT4) {
+            return `vec4.${op}(vec4.create(), vec4.fromValues(${lhs}, ${lhs}, ${lhs}, ${lhs}), ${rhs})`;
+          }
+        } else if (typ === FLOAT3X3) {
+          if (typL === FLOAT3X3 && typR === FLOAT3X3) {
+            return `mat3.${op}(mat3.create(), ${lhs}, ${rhs})`;
+          } else if (typL === FLOAT3X3 && (typR === FLOAT || typR === INT)) {
+            return `mat3.${op}(mat3.create(), ${lhs}, mat3.fromValues(${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}))`;
+          } else if ((typL === FLOAT || typL === INT) && typR === FLOAT3X3) {
+            return `mat3.${op}(mat3.create(), mat3.fromValues(${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}), ${rhs})`;
+          }
+        } else if (typ === FLOAT4X4) {
+          if (typL === FLOAT4X4 && typR === FLOAT4X4) {
+            return `mat4.${op}(mat4.create(), ${lhs}, ${rhs})`;
+          } else if (typL === FLOAT4X4 && (typR === FLOAT || typR === INT)) {
+            return `mat4.${op}(mat4.create(), ${lhs}, mat4.fromValues(${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}, ${rhs}))`;
+          } else if ((typL === FLOAT || typL === INT) && typR === FLOAT4X4) {
+            return `mat4.${op}(mat4.create(), mat4.fromValues(${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}, ${lhs}), ${rhs})`;
```

Comment:
Perhaps it would be useful, for cases like this, to make a little helper function like `mat4fromScalar` that takes a single number as an argument and stamps out an appropriate matrix?